src/oge/load_data.py

```diff
@@ -116,21 +122,251 @@ def load_cems_data(year):
+    for year in range(earliest_data_year, latest_validated_year + 1):
+        cems_id_year = pd.read_parquet(
+            downloads_folder("pudl/hourly_emissions_epacems.parquet"),
+            filters=[["year", "==", year]],
+            columns=["plant_id_epa", "plant_id_eia", "emissions_unit_id_epa"],
+        ).drop_duplicates()
+        cems_ids.append(cems_id_year)
+        cems_ids = [pd.concat(cems_ids, axis=0).drop_duplicates()]
+
+    cems_ids = (
+        pd.concat(cems_ids, axis=0)
+        .drop_duplicates()
+        .sort_values(by=["plant_id_eia", "emissions_unit_id_epa"])
+    )
+
+    cems_ids = apply_pudl_dtypes(cems_ids)
 
     # update the plant_id_eia column using manual matches
-    cems = update_epa_to_eia_map(cems)
+    cems_ids = update_epa_to_eia_map(cems_ids)
+
+    return cems_ids[["plant_id_eia", "emissions_unit_id_epa"]]
+
+
+def load_complete_eia_generators_for_subplants():
+    """
+    Loads a dataframe that contains a complete list of generators, including their
+    unit ids, prime movers, operating dates, and operating status. This will be used
+    when creating subplant IDs to ensure that a complete set of generators is
+    represented. Because some of these values are incomplete or missing in the pudl data
+    we load these values from the raw downloaded EIA-860 dataset to fill in any gaps.
+
+    """
+    complete_gens = load_pudl_table(
+        "denorm_generators_eia",
+        columns=[
+            "report_date",
+            "plant_id_eia",
+            "generator_id",
+            "unit_id_pudl",
+            "prime_mover_code",
+            "operational_status_code",
+            "generator_operating_date",
+            "generator_retirement_date",
+            "original_planned_generator_operating_date",
+            "current_planned_generator_operating_date",
+        ],
+    )
+
+    # create a column that indicates the earliest year a generator reported data to EIA
+    complete_gens["earliest_report_date"] = complete_gens.groupby(
+        ["plant_id_eia", "generator_id"]
+    )["report_date"].transform("min")
+
+    # drop any data that was reported prior to the earliest year
+    # only keep data for years <= the year
+    # this avoids using potentially preliminary early-release data
+    complete_gens = complete_gens[
+        (complete_gens["report_date"].dt.year >= earliest_data_year)
+        & (complete_gens["report_date"].dt.year <= latest_validated_year)
+    ]
+
+    # for any retired gens, forward fill the most recently available unit_id_pudl to the
+    # most recent available year
+    complete_gens["unit_id_pudl"] = complete_gens.groupby(
+        ["plant_id_eia", "generator_id"]
+    )["unit_id_pudl"].ffill()
+
+    # remove generators that retired prior to the earliest year
+    complete_gens = complete_gens[
+        ~(
+            (complete_gens["operational_status_code"] == "RE")
+            & (complete_gens["generator_retirement_date"].dt.year < earliest_data_year)
```

Comment:
Why not removing all generators that retired before `earliest_validated_year`?

Reply:
The idea is that if we ever expand historical data coverage before 2019, we wouldn't be adding a ton of new subplants to the mapping at that point because they would have already been included by considering earliest_data_year. This keeps the subplant_ids as stable as possible over time and between OGE versions